### PR TITLE
Initial user profile implementation

### PR DESCRIPTION
**Pull request by writeriks**
https://github.com/user-attachments/assets/39ebfe5b-065b-4990-9090-adb312adcfd9

https://github.com/user-attachments/assets/2654db57-a21f-4dd0-ad08-54bb75275b0d

### CODE REVIEW

File: src/app/api/user/get-user-posts-by-like/route.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+    const currentUser = await prisma.user.findUnique({
+      where: {
+        firebaseUserId: decodedToken.uid,
+      },
+    });
+
+    if (!currentUser) {
+      return createResponse(ResponseStatus.UNAUTHORIZED);
+    }
```
Comment (Ahmad-Alaziz) [inline src/app/api/user/get-user-posts-by-like/route.ts:25]: this logic seems to be repeated in every api we have, probably it should be extracted.

File: src/components/follow-button/follow-button.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  // Do not render the button if `isUserFollowedData` is undefined
+  if (isUserFollowed === undefined) {
+    return null;
+  }
+
+  // Determine button label
+  const buttonLabel = isUserFollowed ? t('unfollow') : t('follow');
+
+  return (
+    <Button onClick={toggleUserFollow} className='rounded-2xl bg-primary h-8 font-bold'>
```
Comment (Ahmad-Alaziz) [inline src/components/follow-button/follow-button.tsx:84]: can we add some different styling depending on if the button is follow vs unfollow, for better ux. for instance:
Screenshot 2024-12-15 at 18 11 03
Screenshot 2024-12-15 at 18 11 00

**Review by Ahmad-Alaziz — CHANGES_REQUESTED** [review]

File: src/app/[locale]/users/[id]/page.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  // Check if the current user is blocked by the user
+  const isCurrentUserBlockedByUser = await prisma.userBlocks.findFirst({
+    where: {
+      blockedId: currentUser?.userId,
+      blockerId: user.userId,
+    },
+  });
+
+  if (isCurrentUserBlockedByUser) {
+    notFound();
```
[the remark this answers is not in the document]
Comment (Ahmad-Alaziz) [reply src/app/[locale]/users/[id]/page.tsx:70]: maybe we can add a comment for now and leave this as future todo

[the remark this answers is not in the document]
Comment (writeriks) [reply src/app/[locale]/users/[id]/page.tsx:70]: this is when you are blocked by someone and you shouldnt see his/her profile. Then we should show page not found

File: src/components/follow-button/follow-button.tsx (hunk 2 above)
Comment (writeriks) [reply src/components/follow-button/follow-button.tsx:84]: I can add a TODO for it to style the button. But i am not sure how to style it atm.

File: src/app/api/user/get-user-posts-by-date/route.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      stockPostsByDate.length > 0 ? stockPostsByDate[stockPostsByDate.length - 1].postId : null;
+
+    return createResponse(ResponseStatus.OK, {
+      postsByDate: postsWithLikeAndCommentInfo,
+      lastPostIdByDate: newLastPostIdByDate,
+    });
+  } catch (error) {
+    return createResponse(ResponseStatus.INTERNAL_SERVER_ERROR);
+  }
+}
```
Comment (Ahmad-Alaziz) [inline src/app/api/user/get-user-posts-by-date/route.ts:97]: I thought i added this suggestion already but for some reason i dont see it,

Can we combine this logic, for example:

```suggestion
     const posts = await prisma.post.findMany({
    where: { userId: user.userId },
    orderBy: { createdAt: 'desc' },
    take: pageSize,
    cursor: lastPostId ? { postId: lastPostId } : undefined,
    skip: lastPostId ? 1 : 0,
    include: {
      _count: { select: { likedBy: true, comments: true } },
      likedBy: { where: { userId: currentUser.userId }, select: { postId: true } },
    },
  });

  const postsWithInfo = posts.map(post => ({
    ...post,
    likedByCurrentUser: post.likedBy.length > 0,
    likeCount: post._count.likedBy,
    commentCount: post._count.comments,
  }));

  const newLastPostId = posts.length > 0 ? posts[posts.length - 1].postId : null;

  return createResponse(ResponseStatus.OK, {
    postsByDate: postsWithInfo,
    lastPostIdByDate: newLastPostId,
  });
```

File: src/app/api/user/get-user-posts-by-like/route.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      userPostsByLike.length > 0 ? userPostsByLike[userPostsByLike.length - 1].postId : null;
+
+    return createResponse(ResponseStatus.OK, {
+      postsByLike: postsWithLikeAndCommentInfo,
+      lastPostIdByLike: newLastPostIdByLike,
+    });
+  } catch (error) {
+    return createResponse(ResponseStatus.INTERNAL_SERVER_ERROR);
+  }
+}
```
Comment (Ahmad-Alaziz) [inline src/app/api/user/get-user-posts-by-like/route.ts:99]: Same here:

```suggestion
  // Fetch posts sorted by like counts
  const posts = await prisma.post.findMany({
    where: { userId: user.userId },
    orderBy: {
      likedBy: {
        _count: 'desc',
      },
    },
    take: pageSize,
    cursor: lastPostId ? { postId: lastPostId } : undefined,
    skip: lastPostId ? 1 : 0,
    include: {
      _count: { select: { likedBy: true, comments: true } },
      likedBy: { where: { userId: currentUser.userId }, select: { postId: true } },
    },
  });

  const postsWithInfo = posts.map(post => ({
    ...post,
    likedByCurrentUser: post.likedBy.length > 0,
    likeCount: post._count.likedBy,
    commentCount: post._count.comments,
  }));

  const newLastPostId = posts.length > 0 ? posts[posts.length - 1].postId : null;

  return createResponse(ResponseStatus.OK, {
    postsByLike: postsWithInfo,
    lastPostIdByLike: newLastPostId,
  });
}
```

File: src/app/api/user/get-user-posts-by-like/route.ts (hunk 1 above)
Comment (writeriks) [reply src/app/api/user/get-user-posts-by-like/route.ts:25]: here is the task 
https://www.notion.so/Refactor-user-validation-checks-in-route-files-15d9de9c2cb980e59fb6eabdfadb346b?pvs=4

File: src/components/follow-button/follow-button.tsx (hunk 2 above)
Comment (Ahmad-Alaziz) [reply src/components/follow-button/follow-button.tsx:84]: maybe for now something simple like changing bg color, maybe bg-primary to bg-background?

File: src/app/api/user/get-user-posts-by-like/route.ts (hunk 5 above)
Comment (writeriks) [reply src/app/api/user/get-user-posts-by-like/route.ts]: done

File: src/app/api/user/get-user-posts-by-date/route.ts (hunk 4 above)
Comment (writeriks) [reply src/app/api/user/get-user-posts-by-date/route.ts]: done

**Review by Ahmad-Alaziz — APPROVED** [review]
awesome 🚀

**Pull request merged**